File: python/thehallucinatedlab/tools/tokenize.py
"""Count tokens, so a corpus can be budgeted before it is embedded.

This is a measuring instrument, not a stage of the pipeline. An
embedding model tokenizes internally with its own vocabulary -- BGE-M3
is XLM-RoBERTa, SentencePiece, roughly 250k pieces -- so there is no
such thing as handing it a pre-tokenized file. A tiktoken id stream fed
to BGE-M3 is not "already tokenized", it is a different alphabet.

What a tokenizer is genuinely for here is two questions:

    how big should a chunk be, so it fits the model's context?
    what will embedding this corpus cost?

``chunk`` imports :func:`counter_for` to answer the first. The tool
itself answers the second.

Three counters, and the difference matters:

    bge       exact for this pipeline. Downloads ~17MB -- the tokenizer,
              not the 2.3GB model.
    openai    tiktoken, for costing against a different model. Wrong
              for BGE-M3 by construction.
    estimate  no download, no dependency, deliberately high. A chunk
              sized by an under-count overflows the real tokenizer at
              embed time and the tail is silently dropped, so the
              heuristic errs the only safe direction.
"""

from __future__ import annotations

import json
import os
import re
from collections.abc import Callable, Iterable, Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Union

from ..deps import require
from ..errors import InvalidArgument
from ..registry import registry

Source = Union[str, "os.PathLike[str]", bytes, bytearray, Any]

# One token per character for scripts that do not use spaces. Han, kana,
# Hangul: SentencePiece rarely merges these below one piece per glyph,
# and often produces more.
_DENSE = (
    "぀-ヿ"  # kana
    "㐀-䶿"  # CJK extension A
    "一-鿿"  # CJK unified
    "가-힯"  # Hangul
)
_LATIN = "A-Za-zÀ-ɏ"

_PIECE = re.compile(f"[{_DENSE}]|[{_LATIN}]+|\\d+|[^\\s{_LATIN}0-9{_DENSE}]")

# Applied to the heuristic's raw count: +15%. Under-counting is the
# expensive mistake -- it produces chunks that overflow at embed time and
# lose their tail silently -- so the estimate is biased high rather than
# centred. Kept as a fraction so both runtimes can do integer arithmetic.
_SAFETY_NUM = 115
_SAFETY_DEN = 100

_cache: dict[str, Callable[[str], int]] = {}


def estimate_tokens(text: str) -> int:
    """Count tokens without a tokenizer, erring high.

    Mirrors estimateTokens() in toolkit.js exactly. The browser has no
    way to run BGE-M3's tokenizer without a 17MB download, so it chunks
    against this and records that it did -- and ``embed`` re-checks with
    the real thing rather than trusting it.
    """
    body = str(text or "")
    if not body.strip():
        return 0

    total = 0
    for piece in _PIECE.findall(body):
        if re.fullmatch(f"[{_LATIN}]+", piece):
            # ~4 characters per piece for Latin script, rounded up.
            total += max(1, -(-len(piece) // 4))
        elif piece.isdigit():
            # Digits fragment far more than letters; most vocabularies
            # carry only short numeric pieces.
            total += max(1, -(-len(piece) // 2))
        else:
            total += 1
    # Integer arithmetic, not float: ceil(total * 1.15) computed through
    # a float rounds differently in JavaScript at some magnitudes, and
    # the two runtimes must agree exactly or the shared fixtures split.
    return (total * _SAFETY_NUM + _SAFETY_DEN - 1) // _SAFETY_DEN


def _bge_counter() -> Callable[[str], int]:
    tokenizers = require("tokenizers", extra="chunk", purpose="counting BGE-M3 tokens")
    encoder = tokenizers.Tokenizer.from_pretrained("BAAI/bge-m3")

    def count_bge(text: str) -> int:
        return len(encoder.encode(text, add_special_tokens=True).ids)

    return count_bge


def _openai_counter() -> Callable[[str], int]:
    tiktoken = require("tiktoken", extra="chunk", purpose="counting OpenAI tokens")
    encoding = tiktoken.get_encoding("cl100k_base")

    def count_openai(text: str) -> int:
        return len(encoding.encode(text))

    return count_openai


_BUILDERS: dict[str, Callable[[], Callable[[str], int]]] = {
    "bge": _bge_counter,
    "openai": _openai_counter,
    "estimate": lambda: estimate_tokens,
}


def counter_for(tokenizer: str = "bge") -> Callable[[str], int]:
    """A function that counts tokens the way ``tokenizer`` does.

    Cached, because building the BGE counter downloads and parses a
    17MB vocabulary and chunking calls it once per candidate boundary.
    """
    if tokenizer not in _BUILDERS:
        raise InvalidArgument(
            f"Unknown tokenizer {tokenizer!r}. Known: {', '.join(_BUILDERS)}."
        )
    if tokenizer not in _cache:
        _cache[tokenizer] = _BUILDERS[tokenizer]()
    return _cache[tokenizer]


def is_exact(tokenizer: str) -> bool:
    """Whether this counter matches what the embedding model will do."""
    return tokenizer == "bge"


@dataclass
class TokenReport:
    """What a corpus costs, and what will not fit.

    ``over_limit`` is the number that matters. A piece longer than the
    model's context is not an error at embed time -- it is silently
    truncated, and the tail simply never reaches the vector. Finding
    that here is the whole point of the tool.
    """

    tokenizer: str
    exact: bool
    limit: int
    pieces: int
    total: int
    smallest: int
    largest: int
    mean: float
    median: int
    p95: int
    over_limit: int
    empty: int

    def __str__(self) -> str:
        if self.exact:
            how = "exact"
        elif self.tokenizer == "estimate":
            how = "estimated"
        else:
            how = f"{self.tokenizer}, approximate for bge"
        tail = f", {self.over_limit} over {self.limit}" if self.over_limit else ""
        return (
            f"{self.total:,} tokens across {self.pieces:,} piece(s) "
            f"[{how}]; median {self.median}, largest {self.largest}{tail}"
        )

    def as_dict(self) -> dict[str, Any]:
        return {
            "tokenizer": self.tokenizer,
            "exact": self.exact,
            "limit": self.limit,
            "pieces": self.pieces,
            "total": self.total,
            "smallest": self.smallest,
            "largest": self.largest,
            "mean": self.mean,
            "median": self.median,
            "p95": self.p95,
            "over_limit": self.over_limit,
            "empty": self.empty,
        }


def _percentile(ordered: Sequence[int], fraction: float) -> int:
    """Nearest-rank percentile. No numpy for a one-line statistic."""
    if not ordered:
        return 0
    index = max(0, min(len(ordered) - 1, int(round(fraction * (len(ordered) - 1)))))
    return ordered[index]


def analyze(
    texts: Iterable[str],
    tokenizer: str = "bge",
    limit: int = 8192,
) -> TokenReport:
    """Measure an iterable of pieces."""
    count = counter_for(tokenizer)
    counts = [count(text) for text in texts]
    ordered = sorted(counts)
    non_empty = [c for c in counts if c]

    return TokenReport(
        tokenizer=tokenizer,
        exact=is_exact(tokenizer),
        limit=limit,
        pieces=len(counts),
        total=sum(counts),
        smallest=ordered[0] if ordered else 0,
        largest=ordered[-1] if ordered else 0,
        mean=round(sum(counts) / len(counts), 1) if counts else 0.0,
        median=_percentile(ordered, 0.5),
        p95=_percentile(ordered, 0.95),
        over_limit=sum(1 for c in counts if c > limit),
        empty=len(counts) - len(non_empty),
    )


def _pieces_from(payload: bytes, name: str) -> list[str]:
    """One piece per JSONL record, or the whole document otherwise.

    A .jsonl input is assumed to be chunk's output, so each record is
    measured separately -- which is the only way "how many chunks
    overflow" is a meaningful question.
    """
    text = re.sub(r"\r\n?", "\n", payload.decode("utf-8", errors="replace"))
    if not name.lower().endswith(".jsonl"):
        return [text]

    pieces: list[str] = []
    for number, line in enumerate(text.splitlines(), start=1):
        if not line.strip():
            continue
        try:
            record = json.loads(line)
        except json.JSONDecodeError as err:
            raise InvalidArgument(f"{name} line {number} is not valid JSON: {err}") from err
        if not isinstance(record, dict) or "text" not in record:
            raise InvalidArgument(f"{name} line {number} has no 'text' field.")
        pieces.append(str(record["text"]))
    return pieces


def tokenize(
    source: Source,
    *,
    tokenizer: str | None = None,
    limit: int | None = None,
    filename: str | None = None,
) -> TokenReport:
    """Count the tokens in ``source``.

    Args:
        source: Path to a document, raw bytes, a binary file object, or
            a list of strings to measure directly.
        tokenizer: ``"bge"`` (default, exact for this pipeline),
            ``"openai"``, or ``"estimate"``.
        limit: Context window to measure against. Default 8192.
        filename: Name to attribute bytes to, so a .jsonl input is
            recognised and measured per record.

    Returns:
        A :class:`TokenReport`.

    Raises:
        InvalidArgument: an argument is outside what the manifest allows,
            or a JSONL line is malformed.
        DependencyMissing: the chosen tokenizer needs the ``chunk`` extra.

    Example:
        >>> report = tokenize(["hello there"], tokenizer="estimate")
        >>> report.pieces
        1
    """
    args = registry.validate("tokenize", tokenizer=tokenizer, limit=limit)

    if isinstance(source, list | tuple):
        pieces = [str(item) for item in source]
    else:
        if isinstance(source, bytes | bytearray):
            payload, name = bytes(source), filename or ""
        elif hasattr(source, "read"):
            raw = source.read()
            if not isinstance(raw, bytes):
                raise InvalidArgument("A file object must be opened in binary mode.")
            payload, name = raw, filename or ""
        else:
            path = Path(os.fspath(source))
            if not path.is_file():
                raise InvalidArgument(f"No such file: {path}")
            payload, name = path.read_bytes(), filename or path.name
        pieces = _pieces_from(payload, name)

    return analyze(pieces, args["tokenizer"], args["limit"])


registry.register("tokenize", tokenize)

File: tokenize.js
/* ============================================================
   tokenize.js — page glue for tokenize.html.

   Answers "how big is this, and what will it cost to embed" — not
   "convert this document into tokens". That distinction is the whole
   point of the page. An embedding model tokenizes internally with its
   own vocabulary; BGE-M3 is XLM-RoBERTa with roughly 250k SentencePiece
   pieces, so there is no file you can hand it that is "already
   tokenized". A tiktoken id stream is not a head start, it is a
   different alphabet.

   What a token count is genuinely for is deciding how to chunk, and
   knowing what a corpus costs before paying for it.

   The browser counts with toolkit.estimateTokens, a heuristic, because
   BGE-M3's real vocabulary is a ~17MB download. The page says so rather
   than presenting an estimate as a measurement. For exact figures the
   Python package loads the real tokenizer — which is still only the
   tokenizer, not the 2.3GB model.

   Failure model:
     - Spec fails to load -> drop zone disabled, panel says why.
     - Malformed JSONL -> names the line rather than the file.
     - Everything else is arithmetic and cannot fail.
   ============================================================ */
(function () {
  'use strict';

  var toolkit = window.THL && window.THL.toolkit;

  /* @pure-start — no DOM, no network, no module state. */

  /* One piece per JSONL record, or the whole document otherwise.

     A .jsonl input is assumed to be chunk's output, so each record is
     measured separately — which is the only way "how many pieces
     overflow the context window" is a meaningful question at all. */
  function piecesFrom(text, name) {
    var body = String(text || '').replace(/\r\n?/g, '\n');
    if (!/\.jsonl$/i.test(String(name || ''))) return [body];

    var pieces = [];
    var lines = body.split('\n');
    for (var i = 0; i < lines.length; i++) {
      if (!lines[i].trim()) continue;
      var record;
      try {
        record = JSON.parse(lines[i]);
      } catch (err) {
        throw new Error('Line ' + (i + 1) + ' is not valid JSON.');
      }
      if (!record || typeof record !== 'object' || record.text === undefined) {
        throw new Error('Line ' + (i + 1) + ' has no "text" field.');
      }
      pieces.push(String(record.text));
    }
    return pieces;
  }

  /* Nearest-rank percentile. No dependency for a one-line statistic. */
  function percentile(ordered, fraction) {
    if (!ordered.length) return 0;
    var index = Math.round(fraction * (ordered.length - 1));
    return ordered[Math.max(0, Math.min(ordered.length - 1, index))];
  }

  /* The distribution, not just the total. `overLimit` is the number that
     matters: a piece longer than the model's context is not an error at
     embed time, it is silently truncated, and its tail never reaches the
     vector. Finding that here is the point of the tool. */
  function analyze(pieces, limit, count) {
    var counts = [];
    for (var i = 0; i < pieces.length; i++) counts.push(count(pieces[i]));

    var ordered = counts.slice().sort(function (a, b) { return a - b; });
    var total = 0;
    var over = 0;
    var empty = 0;
    for (var j = 0; j < counts.length; j++) {
      total += counts[j];
      if (counts[j] > limit) over++;
      if (!counts[j]) empty++;
    }

    return {
      pieces: counts.length,
      total: total,
      smallest: ordered.length ? ordered[0] : 0,
      largest: ordered.length ? ordered[ordered.length - 1] : 0,
      mean: counts.length ? Math.round((total / counts.length) * 10) / 10 : 0,
      median: percentile(ordered, 0.5),
      p95: percentile(ordered, 0.95),
      overLimit: over,
      empty: empty,
      limit: limit
    };
  }

  /* A histogram of where the pieces fall, for the bar chart. Fixed
     buckets rather than computed ones, so two runs of the same corpus
     are comparable by eye. */
  function histogram(counts, limit) {
    var edges = [0, 64, 128, 256, 512, 1024, 2048, 4096, limit];
    var buckets = [];
    for (var b = 0; b < edges.length - 1; b++) {
      if (edges[b] >= limit) break;
      buckets.push({ from: edges[b], to: Math.min(edges[b + 1], limit), n: 0 });
    }
    buckets.push({ from: limit, to: Infinity, n: 0 });

    for (var i = 0; i < counts.length; i++) {
      for (var k = 0; k < buckets.length; k++) {
        if (counts[i] > buckets[k].from && counts[i] <= buckets[k].to) { buckets[k].n++; break; }
        if (k === buckets.length - 1 && counts[i] > limit) buckets[k].n++;
      }
    }
    return buckets;
  }

  /* @pure-end */

  /* ---- DOM glue ---- */

  var manifest = null;
  var tool = null;
  var sourceFile = null;
  var busy = false;
  var el = {};

  function setStatus(message, kind) {
    if (!el.status) return;
    el.status.textContent = message || '';
    el.status.className = 'tool-status' + (kind ? ' is-' + kind : '');
  }

  function readAsText(file) {
    return new Promise(function (resolve, reject) {
      var reader = new FileReader();
      reader.onload = function () { resolve(String(reader.result)); };
      reader.onerror = function () { reject(new Error('That file could not be read.')); };
      reader.readAsText(file);
    });
  }

  function row(label, value, note) {
    var li = document.createElement('li');
    var name = document.createElement('span');
    name.textContent = label;
    li.appendChild(name);
    li.appendChild(document.createTextNode(' ' + value + (note ? ' — ' + note : '')));
    return li;
  }

  function render(stats, counts) {
    while (el.facts.firstChild) el.facts.removeChild(el.facts.firstChild);
    el.facts.appendChild(row('pieces', stats.pieces.toLocaleString()));
    el.facts.appendChild(row('total', stats.total.toLocaleString() + ' tokens'));
    el.facts.appendChild(row('median', String(stats.median)));
    el.facts.appendChild(row('largest', String(stats.largest)));
    el.facts.appendChild(row('95th percentile', String(stats.p95)));
    el.facts.appendChild(row(
      'over ' + stats.limit,
      String(stats.overLimit),
      stats.overLimit ? 'these would be truncated at embed time' : 'nothing would be truncated'
    ));
    if (stats.empty) el.facts.appendChild(row('empty', String(stats.empty)));

    /* A text histogram rather than a chart: it needs no library, copies
       into a terminal, and reads the same in both themes. */
    var lines = [];
    var buckets = histogram(counts, stats.limit);
    var widest = 0;
    for (var i = 0; i < buckets.length; i++) widest = Math.max(widest, buckets[i].n);
    for (var b = 0; b < buckets.length; b++) {
      var label = buckets[b].to === Infinity
        ? '> ' + stats.limit
        : buckets[b].from + '-' + buckets[b].to;
      var width = widest ? Math.round((buckets[b].n / widest) * 32) : 0;
      lines.push(
        label.padStart(11) + '  ' +
        new Array(width + 1).join('#').padEnd(32) + ' ' + buckets[b].n
      );
    }
    el.preview.textContent = lines.join('\n');
    el.result.hidden = false;
  }

  function run() {
    if (busy || !sourceFile) return;
    var validated = toolkit.validateArgs({
      tokenizer: 'estimate',
      limit: parseInt(el.limit.value, 10)
    }, tool);
    if (!validated.ok) { setStatus(validated.errors.join(' '), 'error'); return; }

    busy = true;
    setStatus('Counting…');

    readAsText(sourceFile).then(function (text) {
      var pieces = piecesFrom(text, sourceFile.name);
      var counts = [];
      for (var i = 0; i < pieces.length; i++) counts.push(toolkit.estimateTokens(pieces[i]));
      render(analyze(pieces, validated.args.limit, toolkit.estimateTokens), counts);
      setStatus('Estimated in this tab. For exact BGE-M3 counts: thl tool tokenize ' +
        sourceFile.name, 'ok');
    }).catch(function (err) {
      setStatus(err.message || 'Counting failed.', 'error');
    }).then(function () {
      busy = false;
    });
  }

  function acceptFile(file) {
    if (!file) return;
    sourceFile = file;
    el.dropTitle.textContent = file.name;
    el.dropHint.textContent = /\.jsonl$/i.test(file.name)
      ? 'JSONL — each record will be counted separately'
      : Math.max(1, Math.round(file.size / 1024)) + ' KB';
    el.run.disabled = false;
    setStatus('');
  }

  document.addEventListener('DOMContentLoaded', function () {
    el = {
      drop: document.getElementById('drop-zone'),
      dropTitle: document.getElementById('drop-title'),
      dropHint: document.getElementById('drop-hint'),
      pick: document.getElementById('pick-file'),
      file: document.getElementById('file-input'),
      limit: document.getElementById('opt-limit'),
      run: document.getElementById('tokenize-btn'),
      status: document.getElementById('tool-status'),
      result: document.getElementById('tool-result'),
      facts: document.getElementById('result-facts'),
      preview: document.getElementById('result-preview'),
      args: document.getElementById('tool-args')
    };

    if (!toolkit) { setStatus('The tool runtime failed to load.', 'error'); return; }

    toolkit.loadManifest().then(function (loaded) {
      manifest = loaded;
      tool = toolkit.findTool(manifest, 'tokenize');
      if (!tool) throw new Error('tokenize is missing from the tool spec.');

      var limit = tool.params.filter(function (p) { return p.name === 'limit'; })[0];
      el.limit.min = limit.min;
      el.limit.max = limit.max;
      el.limit.value = limit.default;

      if (el.args) toolkit.renderParamTable(tool, el.args);
      el.drop.classList.remove('is-disabled');
      setStatus('');
    }).catch(function (err) {
      setStatus(err.message || 'The tool spec failed to load.', 'error');
    });

    el.pick.addEventListener('click', function () { el.file.click(); });
    el.file.addEventListener('change', function () { acceptFile(el.file.files[0]); });
    el.run.addEventListener('click', run);

    ['dragenter', 'dragover'].forEach(function (name) {
      el.drop.addEventListener(name, function (event) {
        event.preventDefault();
        el.drop.classList.add('is-over');
      });
    });
    ['dragleave', 'drop'].forEach(function (name) {
      el.drop.addEventListener(name, function (event) {
        event.preventDefault();
        el.drop.classList.remove('is-over');
      });
    });
    el.drop.addEventListener('drop', function (event) {
      acceptFile(event.dataTransfer.files[0]);
    });
  });

})();
